server/forge/bitbucketdatacenter/convert.go

```diff
@@ -0,0 +1,150 @@
+	"fmt"
+	"strings"
+	"time"
+
+	bb "github.com/neticdk/go-bitbucket/bitbucket"
+
+	"go.woodpecker-ci.org/woodpecker/v2/server/model"
+)
+
+func convertStatus(status model.StatusValue) bb.BuildStatusState {
+	switch status {
+	case model.StatusPending, model.StatusRunning:
+		return bb.BuildStatusStateInProgress
+	case model.StatusSuccess:
+		return bb.BuildStatusStateSuccessful
+	default:
+		return bb.BuildStatusStateFailed
+	}
+}
+
+func convertID(id uint64) model.ForgeRemoteID {
+	return model.ForgeRemoteID(fmt.Sprintf("%d", id))
+}
+
+func convertRepo(from *bb.Repository, perm *model.Perm, branch string) *model.Repo {
+	r := &model.Repo{
+		ForgeRemoteID: convertID(from.ID),
+		Name:          from.Slug,
+		Owner:         from.Project.Key,
+		Branch:        branch,
+		SCMKind:       model.RepoGit,
+		IsSCMPrivate:  true, // Since we have to use Netrc it has to always be private :/ TODO: Is this really true?
+		FullName:      fmt.Sprintf("%s/%s", from.Project.Key, from.Slug),
+		Perm:          perm,
+		PREnabled:     true,
+	}
+
+	for _, l := range from.Links["clone"] {
+		if l.Name == "http" {
+			r.Clone = l.Href
+		}
+	}
+
+	if l, ok := from.Links["self"]; ok && len(l) > 0 {
+		r.ForgeURL = l[0].Href
+	}
+
+	return r
+}
+
+func convertRepositoryPushEvent(ev *bb.RepositoryPushEvent, baseURL string) *model.Pipeline {
+	if len(ev.Changes) == 0 {
+		return nil
+	}
+	change := ev.Changes[0]
+	if change.ToHash == "0000000000000000000000000000000000000000" {
+		// No ToHash present - could be "DELETE"
+		return nil
+	}
+	if change.Type == bb.RepositoryPushEventChangeTypeDelete {
+		return nil
+	}
+
+	pipeline := &model.Pipeline{
+		Commit:    change.ToHash,
+		Branch:    change.Ref.DisplayID,
+		Message:   "",
+		Avatar:    bitbucketAvatarURL(baseURL, ev.Actor.Slug),
+		Author:    authorLabel(ev.Actor.Name),
+		Email:     ev.Actor.Email,
+		Timestamp: time.Time(ev.Date).UTC().Unix(),
+		Ref:       ev.Changes[0].RefId,
+		ForgeURL:  fmt.Sprintf("%s/projects/%s/repos/%s/commits/%s", baseURL, ev.Repository.Project.Key, ev.Repository.Slug, change.ToHash),
+	}
+
+	if strings.HasPrefix(ev.Changes[0].RefId, "refs/tags/") {
+		pipeline.Event = model.EventTag
+	} else {
+		pipeline.Event = model.EventPush
+	}
+
+	return pipeline
+}
+
+func convertPullRequestEvent(ev *bb.PullRequestEvent, baseURL string) *model.Pipeline {
+	pipeline := &model.Pipeline{
+		Commit:    ev.PullRequest.Source.Latest,
+		Branch:    ev.PullRequest.Source.DisplayID,
+		Title:     ev.PullRequest.Title,
+		Message:   "",
+		Avatar:    bitbucketAvatarURL(baseURL, ev.Actor.Slug),
+		Author:    authorLabel(ev.Actor.Name),
+		Email:     ev.Actor.Email,
+		Timestamp: time.Time(ev.Date).UTC().Unix(),
+		Ref:       fmt.Sprintf("refs/pull-requests/%d/from", ev.PullRequest.ID),
+		ForgeURL:  fmt.Sprintf("%s/projects/%s/repos/%s/commits/%s", baseURL, ev.PullRequest.Source.Repository.Project.Key, ev.PullRequest.Source.Repository.Slug, ev.PullRequest.Source.Latest),
+		Event:     model.EventPull,
```

Comment:
Do you know if this hook also gets called when closing a PR? We recently introduced a `pull_request_closed` event in case.

Reply:
@anbraten  That is a good question - it should be possible to get an event from Bitbucket when pull-request is either deleted, merged or declined - that would be the Bitbucket "termination" states. I do not think the webhook is registered with these now but we could add that. You think we should include this as part of this or add it in a smaller PR afterwards?

Reply:
It does - the webhook is currently registeret with these events (bitbucketdatacenter.go:404)
```go
Events: []bb.EventKey{bb.EventKeyRepoRefsChanged, bb.EventKeyPullRequestFrom},
```

I think I checked with the old Bitbucket plugin as well as the current Bitbucket cloud plugin which events they react to - there is s list of the events in the Bitbucket client: https://github.com/neticdk/go-bitbucket/blob/main/bitbucket/events.go#L55

Reply:
If we don't add the event now users will later have to click repair on all their repos to update the webhook events. Just wanted to mention that. Both ways would be fine for me.

Reply:
that is true 🤔 

shouldn't be that big a change, I'll see if I can find a few hours to put it in